### PR TITLE
[wasm] Build static components; include hot_reload in runtime

### PR DESCRIPTION
**Pull request by lambdageek**
Workaround until https://github.com/dotnet/runtime/issues/54565 is fixed
Build the runtime always with support for hot_reload, and without diagnostics_tracing

### CODE REVIEW

File: src/mono/wasm/Makefile [hunk 1]
```diff
@@ -48,12 +48,19 @@ provision-wasm: .stamp-wasm-install-and-select-$(EMSCRIPTEN_VERSION)
...
+	$(MONO_BIN_DIR)/libmono-component-diagnostics_tracing-stub-static.a
+
 MONO_OBJ_DIR=$(OBJDIR)/mono/Browser.wasm.$(CONFIG)
 MONO_INCLUDE_DIR=$(MONO_BIN_DIR)/include/mono-2.0
 BUILDS_OBJ_DIR=$(MONO_OBJ_DIR)/wasm
 MONO_LIBS = \
 	$(MONO_BIN_DIR)/libmono-ee-interp.a \
 	$(MONO_BIN_DIR)/libmonosgen-2.0.a \
+	$(MONO_COMPONENT_LIBS) \
+	$(MONO_BIN_DIR)/libmonosgen-2.0.a \
```
Comment (lambdageek) [inline src/mono/wasm/Makefile:63]: libmonosgen-2.0 is here twice intentionally - the components depend on the runtime and the runtime depends on the components.

**Comment by lambdageek** [comment]
This also needs to go into .NET 6 Preview 6

File: src/mono/wasm/Makefile (hunk 1 above)
Comment (radical) [reply src/mono/wasm/Makefile:63]: add a comment mentioning this?

**Comment by radical** [comment]
Does this enable, or fix something?

**Comment by lambdageek** [comment]
> Does this enable, or fix something?

Yea it fixes hot reload on .NET 6 P6

**Comment by radical** [comment]
> > Does this enable, or fix something?
> 
> Yea it fixes hot reload on .NET 6 P6

Is there a test that we can enable with this, so we get validation too?

**Comment by lambdageek** [comment]
@radical I added a new functional test for the browser.  didn't try very hard to override all the build lane stuff, but at least for a default `./build.sh --os browser -c Release` and running the test with `dotnet.sh build /p:TargetOS=browser /p:TargetArchitecture=wasm /t:Test` works for me.

**Comment by radical** [comment]
wasm failure is https://github.com/dotnet/runtime/issues/53957

File: src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@ namespace System.Reflection.Metadata
     public class ApplyUpdateTest
     {
         [Fact]
+	[ActiveIssue("https://github.com/dotnet/runtime/issues/54617", typeof(PlatformDetection), nameof(PlatformDetection.IsBrowser), nameof(PlatformDetection.IsMonoAOT))]
```
Comment (radical) [inline src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:21]: nit: spaces instead of tab

File: src/mono/wasm/wasm.proj [hunk 3]
```diff
@@ -189,13 +189,16 @@
...
+      <MonoComponentLibs Include="$(MonoArtifactsPath)libmono-component-hot_reload-static.a;
+                                  $(MonoArtifactsPath)libmono-component-diagnostics_tracing-stub-static.a" />
       <MonoLibFiles Include="$(MonoArtifactsPath)libmono-ee-interp.a;
                              $(MonoArtifactsPath)libmonosgen-2.0.a;
                              $(MonoArtifactsPath)libmono-ilgen.a;
                              $(MonoArtifactsPath)libmono-icall-table.a;
                              $(NativeBinDir)libSystem.Native.a;
                              $(NativeBinDir)libSystem.IO.Compression.Native.a" />
       <MonoLibFiles Include="@(ICULibNativeFiles)" />
+      <MonoLibFiles Include="@(MonoComponentLibs)" />
```
Comment (radical) [inline src/mono/wasm/wasm.proj:201]:
```suggestion
      <MonoLibFiles Include="@(MonoComponentLibs)" />
```

File: src/tests/FunctionalTests/WebAssembly/Browser/HotReload/ApplyUpdateReferencedAssembly/ApplyUpdateReferencedAssembly.csproj [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    <Compile Include="MethodBody1.cs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <!-- This package from https://github.com/dotnet/hotreload-utils provides
+	 targets that read the json delta script and generates deltas based on the baseline assembly and the modified sources.
+
+	 Projects must define the DeltaScript property that specifies the (relative) path to the json script.
+	 Deltas will be emitted next to the output assembly.  Deltas will be copied when the current
+	 project is referenced from other other projects.
```
Comment (radical) [inline src/tests/FunctionalTests/WebAssembly/Browser/HotReload/ApplyUpdateReferencedAssembly/ApplyUpdateReferencedAssembly.csproj:26]:
```suggestion
     targets that read the json delta script and generates deltas based on the baseline assembly and the modified sources.

     Projects must define the DeltaScript property that specifies the (relative) path to the json script.
     Deltas will be emitted next to the output assembly.  Deltas will be copied when the current
     project is referenced from other other projects.
```

File: src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+          Include="%(ResolvedFileToPublish.FullPath)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="IncludeDeltasInWasmBundle"
+	  BeforeTargets="PrepareForWasmBuildApp"
+	  Condition="'$(TargetOS)' == 'Browser'">
+    <ItemGroup>
+      <!-- FIXME: this belongs in eng/testing/tests.wasm.targets -->
+      <!-- FIXME: Can we do something on the Content items in the referenced projects themselves to get this for free? -->
```
Comment (radical) [inline src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj:42]: soon ;)

File: src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj [hunk 6]
```diff
@@ -0,0 +1,52 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (radical) [inline src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj:1]: tabs -> spaces for this whole file

File: src/tests/FunctionalTests/WebAssembly/Browser/HotReload/Program.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            ApplyUpdate(assm);
+
+            r = ApplyUpdateReferencedAssembly.MethodBody1.StaticMethod1();
+            if ("NEWEST STRING" != r)
+                return failure;
+
+            return success;
+        }
+
+
```
Comment (radical) [inline src/tests/FunctionalTests/WebAssembly/Browser/HotReload/Program.cs:55]:
```suggestion
```

**Review by radical — APPROVED** [review]
Looks good, just nits, whatever bits I could review. The env vars stuff should be doable in the project file itself, but nothing to block this PR over.

**Pull request merged**